### PR TITLE
Rename and add options to `this-in-template`.

### LINKED ISSUE #148: Disallow `this` in template
**Please describe what the rule should do:**
Disallow `this` in the template

**What category of rule is this? (place an "X" next to just one item)**

[x] Enforces code style
[ ] Warns about a potential error
[ ] Suggests an alternate way of doing something
[ ] Other (please specify:)

**Provide 2-3 code examples that this rule will warn about:**

```html
<template>
     <a :href="this.link">{{this.text}}</a>
</template>
```

```html
<template>
     <div :class="this.$style.container">
          <slot />
     </div>
</template>
```

### PR DESCRIPTION
**Pull request by armano2**
fixes #148

### CODE REVIEW

File: lib/rules/this-in-template.js [hunk 1]
```diff
@@ -0,0 +1,99 @@
...
+    const options = context.options[0] !== 'always' ? 'never' : 'always'
+    let scope = {
+      parent: null,
+      nodes: []
+    }
+
+    function validateNever () {
+      return {
+        'VExpressionContainer ThisExpression' (node) {
+          if (options === 'never') {
```
Comment (mysticatea) [inline lib/rules/this-in-template.js]: I think that it should check whether variables which have the same name as the property name.

```html
<template>
    <div v-for="x of xs">
        {{this.x}} <!-- we cannot remove this `this.`. -->
    </div>
</template>
```

Also, I think that it should check whether the `this` is `MemberExpression#object`.

```html
<template>
    <child :parent="this" /> <!-- we cannot remove this `this`. -->
</template>
```

**Comment by mysticatea** [comment]
Incidentally, I have improved `vue-eslint-parser` as having the relationship between variables and references: https://github.com/mysticatea/vue-eslint-parser/commit/02579df15c7eeb7658c216036895be19818a7bcb.

- `Variable#references` is an array of references which use the variable. If this is empty, it's an unused variable.
- `Reference#variable` is the variable that the reference uses. If this is `null`, it's global variable or the member of the instance.

Though, I'm not sure that it helps this rule.

**Comment by armano2** [comment]
@mysticatea i i like changes with `references` it can simplify this rule, but for now there is infinite loop there: 
https://github.com/mysticatea/vue-eslint-parser/issues/16

**Comment by mysticatea** [comment]
Thank you!

I'm sorry, I have realized one more thing. If property name is a reserved word (e.g. `this.class`), we cannot remove the `this.` since it causes syntax errors. Could you add the check into `"never"`?

**Comment by armano2** [comment]
@mysticatea check for js reserved names added

**Comment by armano2** [comment]
additionality i added support for `this['0aaaa']` and `this['foo-bar bas']` witch should not be warned

**Comment by armano2** [comment]
@mysticatea with latest change to parser, reference variables are no longer there

_Sent from my Xiaomi Redmi 4A using [FastHub](https://play.google.com/store/apps/details?id=com.fastaccess.github)_

**Comment by mysticatea** [comment]
I'm sorry for late.

> additionality i added support for this['0aaaa'] and this['foo-bar bas'] which should not be warned

Good catch!

> with latest change to parser, reference variables are no longer there

I think, in fact, `console.log()` no longer shows `Variable#references` and `Reference#variable` since those are non-enumerable properties. But those are still there. Tests exist at [vue-eslint-parser/test/variables-references.js#L152-L248](https://github.com/mysticatea/vue-eslint-parser/blob/426a9e399e7cca7f8c6a73989af40b69095ff893/test/variables-references.js#L152-L248).

**Comment by michalsnik** [comment]
GJ! I'll release it at the end of this weekend @mysticatea @armano2 . I'm moving to new place, so have quite a lot on my plate lately..

**Pull request merged**